pkg/v1/cli/pluginmanager/manager.go

```diff
@@ -566,3 +589,108 @@ func Clean() error {
 	}
 	return os.RemoveAll(common.DefaultPluginRoot)
 }
+
+// getCLIPluginResourceWithLocalDistroFromPluginDescriptor return cliv1alpha1.CLIPlugin resource from the plugin descriptor
+// Note: This function generates cliv1alpha1.CLIPlugin which contains only single local distribution type artifact for
+// OS-ARCH where user is running the cli
+// This function is only used to create CLIPlugin resource for local plugin installation with legacy directory structure
+func getCLIPluginResourceWithLocalDistroFromPluginDescriptor(pd *cliv1alpha1.PluginDescriptor, pluginBinaryPath string) cliv1alpha1.CLIPlugin {
+	return cliv1alpha1.CLIPlugin{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: pd.Name,
+		},
+		Spec: cliv1alpha1.CLIPluginSpec{
+			Description:        pd.Description,
+			RecommendedVersion: pd.Version,
+			Artifacts: map[string]cliv1alpha1.ArtifactList{
+				pd.Version: []cliv1alpha1.Artifact{
+					{
+						URI:  pluginBinaryPath,
+						Type: common.DistributionTypeLocal,
+						OS:   runtime.GOOS,
+						Arch: runtime.GOARCH,
+					},
+				},
+			},
+		},
+	}
+}
+
+// DiscoverPluginsFromLocalSourceWithLegacyDirectoryStructure returns the available plugins
+// that are discovered from the provided local path
+func DiscoverPluginsFromLocalSourceWithLegacyDirectoryStructure(localPath string) ([]plugin.Discovered, error) {
+	if localPath == "" {
+		return nil, nil
+	}
+
+	// Get the plugin manifest object from manifest.yaml file
+	manifest, err := getPluginManifestResource(filepath.Join(localPath, ManifestFileName))
+	if err != nil {
+		return nil, err
+	}
+
+	var discoveredPlugins []plugin.Discovered
+
+	// Create plugin.Discovered object for all locally available plugin
+	for _, p := range manifest.Plugins {
+		if p.Name == cli.CoreName {
+			continue
+		}
+
+		// Get the plugin descriptor from the plugin.yaml file
+		pd, err := getPluginDescriptorResource(filepath.Join(localPath, p.Name, PluginFileName))
+		if err != nil {
+			return nil, fmt.Errorf("could not unmarshal plugin.yaml: %v", err)
+		}
+
+		absLocalPath, err := filepath.Abs(localPath)
+		if err != nil {
+			return nil, err
+		}
+		// With legacy configuration directory structure creating the pluginBinary path from plugin descriptor
+		// Sample path: cli/<plugin-name>/<plugin-version>/tanzu-<plugin-name>-<os>_<arch>
+		// 				cli/login/v0.14.0/tanzu-login-darwin_amd64
+		pluginBinaryPath := filepath.Join(absLocalPath, p.Name, pd.Version, fmt.Sprintf("tanzu-%s-%s_%s", p.Name, runtime.GOOS, runtime.GOARCH))
```

Comment:
It probably achieves the same effect of failing on a os arch mismatch, but conceptually shouldn't the os/arch be determined by the contents of the directory rather than the system's os/arch? Same for getCLIPluginResourceWithLocalDistroFromPluginDescriptor.
This warrants comment that we are constructing what we 'expect' the local dir to contain.

Another option would be to error out because our expectation does not reflect what is physically in the directory (e.g. someone untar a bundle intended for a difference os/arch) instead of constructing Discovered plugin resources only to fail to use them later.

Reply:
Ack and updated